notebooks/245-llm-chatbot/245-llm-chatbot.ipynb

```diff
@@ -0,0 +1,966 @@
+    "    Params:\n",
+    "      history: dialogue history\n",
+    "    Returns:\n",
+    "      history in text format\n",
+    "    \"\"\"\n",
+    "    text = start_message + \"\".join(\n",
+    "        [\n",
+    "            \"\".join(\n",
+    "                [\n",
+    "                    history_template.format(user=item[0], assistant=item[1])\n",
+    "                ]\n",
+    "            )\n",
+    "            for item in history[:-1]\n",
+    "        ]\n",
+    "    )\n",
+    "    text += \"\".join(\n",
+    "        [\n",
+    "            \"\".join(\n",
+    "                [\n",
+    "                    current_message_template.format(user=history[-1][0], assistant=history[-1][1])\n",
+    "                ]\n",
+    "            )\n",
+    "        ]\n",
+    "    )\n",
+    "    return text\n",
+    "\n",
+    "\n",
+    "\n",
+    "def user(message, history):\n",
+    "    \"\"\"\n",
+    "    callback function for updating user messages in interface on submit button click\n",
+    "    \n",
+    "    Params:\n",
+    "      message: current message\n",
+    "      history: conversation history\n",
+    "    Returns:\n",
+    "      None\n",
+    "    \"\"\"\n",
+    "    # Append the user's message to the conversation history\n",
+    "    return \"\", history + [[message, \"\"]]\n",
+    "\n",
+    "\n",
+    "def bot(history, temperature, top_p, top_k, repetition_penalty, conversation_id):\n",
+    "    \"\"\"\n",
+    "    callback function for running chatbot on submit button click\n",
+    "    \n",
+    "    Params:\n",
+    "      history: conversation history\n",
+    "      temperature:  parameter for control the level of creativity in AI-generated text. \n",
+    "                    By adjusting the `temperature`, you can influence the AI model's probability distribution, making the text more focused or diverse.\n",
+    "      top_p: parameter for control the range of tokens considered by the AI model based on their cumulative probability.\n",
+    "      top_k: parameter for control the range of tokens considered by the AI model based on their cumulative probability, selecting number of tokens with highest probability.\n",
+    "      repetition_penalty: parameter for penalizing tokens based on how frequently they occur in the text.\n",
+    "      conversation_id: unique conversation identifier.\n",
+    "    \n",
+    "    \"\"\"\n",
+    "\n",
+    "    # Construct the input message string for the model by concatenating the current system message and conversation history\n",
+    "    messages = convert_history_to_text(history)\n",
+    "\n",
+    "    # Tokenize the messages string\n",
+    "    input_ids = tok(messages, return_tensors=\"pt\", **tokenizer_kwargs).input_ids\n",
+    "    streamer = TextIteratorStreamer(tok, timeout=10.0, skip_prompt=True, skip_special_tokens=True)\n",
```

Comment:
I noticed if the response does not get generated by 10 seconds an exception will be raised which is not handled. Something worth looking into.

Reply:
However, I tried the code in a python program standalone in a macbook pro 2.4 GHz 8-Core Intel Core i9, 64 GB 2667 MHz DDR4